File: src/main/java/net/volcano/jdacommands/model/command/arguments/interfaces/Codec.java
package net.volcano.jdacommands.model.command.arguments.interfaces;

import net.volcano.jdacommands.model.ParameterData;
import net.volcano.jdacommands.model.command.annotations.argument.Arg;
import net.volcano.jdacommands.model.command.annotations.argument.Optional;
import net.volcano.jdacommands.model.command.arguments.CommandArgument;
import net.volcano.jdautilities.utils.ClassUtil;

import javax.annotation.Nullable;

public abstract class Codec<T> {
	
	protected Class<? extends CommandArgument<T>> argumentClass;
	
	public CommandArgument<T> encodeArgument(ParameterData data) {
		
		var expectedType = ClassUtil.stripWildcard(ClassUtil.getGenericType(getClass()));
		
		if (data.codecType != expectedType) {
			throw new IllegalArgumentException("Invalid codec " + data.codecType.getTypeName() + " used for type " + expectedType.getTypeName() + ".");
		}
		var arg = buildArgument(data);
		
		var ann = data.parameter.getAnnotation(Arg.class);
		if (ann != null) {
			if (!ann.usage().equals("DEFAULT")) {
				arg.setUsage(ann.usage());
			}
		}
		
		arg.setNullable(data.parameter.isAnnotationPresent(Nullable.class));
		
		arg.setOptional(data.parameter.isAnnotationPresent(Optional.class));
		
		arg.setParameter(data.parameter);
		
		arg.setType(data.actualType);
		
		return arg;
		
	}
	
	protected abstract CommandArgument<T> buildArgument(ParameterData data);
	
}

File: src/main/java/net/volcano/jdacommands/model/command/arguments/interfaces/Codec.kt
package net.volcano.jdacommands.model.command.arguments.interfaces

import net.volcano.jdacommands.model.ParameterData
import net.volcano.jdacommands.model.command.annotations.argument.Arg
import net.volcano.jdacommands.model.command.arguments.CommandArgument
import net.volcano.jdautils.utils.ClassUtil

abstract class Codec<T> {

	fun encodeArgument(data: ParameterData): CommandArgument<T> {
		require(data.codecClass === ClassUtil.stripWildcard(ClassUtil.getGenericType(javaClass))) { "Invalid codec used for type." }

		val arg = buildArgument(data)
		val ann = data.parameter.annotations.firstOrNull { it is Arg } as Arg?

		if (arg.usage == null)
			arg.usage = ann?.usage?.let { if (it == "DEFAULT") null else it }

		arg.parameter = data.parameter
		arg.type = data.actualClass

		return arg
	}

	protected abstract fun buildArgument(data: ParameterData): CommandArgument<T>

}